futures-executor/src/thread_pool.rs

```diff
@@ -0,0 +1,366 @@
+            .finish()
+    }
+}
+
+enum Message {
+    Run(Task),
+    Close,
+}
+
+impl ThreadPool {
+    /// Creates a new thread pool with `size` worker threads associated with it.
+    ///
+    /// The returned handle can use `execute` to run work on this thread pool,
+    /// and clones can be made of it to get multiple references to the same
+    /// thread pool.
+    ///
+    /// This is a shortcut for:
+    ///
+    /// ```rust
+    /// # use futures_executor::{ThreadPoolBuilder, ThreadPool};
+    /// #
+    /// # fn new(size: usize) -> ThreadPool {
+    /// ThreadPoolBuilder::new().pool_size(size).create()
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if `size == 0`.
+    pub fn new(size: usize) -> ThreadPool {
+        ThreadPoolBuilder::new().pool_size(size).create()
+    }
+
+    /// Creates a new thread pool with a number of workers equal to the number
+    /// of CPUs on the host.
+    ///
+    /// This is a shortcut for:
+    ///
+    /// ```rust
+    /// # use futures_executor::{ThreadPoolBuilder, ThreadPool};
+    /// #
+    /// # fn new_num_cpus() -> ThreadPool {
+    /// ThreadPoolBuilder::new().create()
+    /// # }
+    /// ```
+    pub fn new_num_cpus() -> ThreadPool {
+        ThreadPoolBuilder::new().create()
+    }
+
+    /// dox
+    pub fn builder() -> ThreadPoolBuilder {
+        ThreadPoolBuilder::new()
+    }
+}
+
+impl Executor for ThreadPool {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>) -> Result<(), SpawnError> {
+        let task = Task {
+            spawn: f,
+            map: LocalMap::new(),
+            wake_handle: Arc::new(WakeHandle {
+                exec: self.clone(),
+                mutex: UnparkMutex::new(),
+            }),
+        };
+        self.state.send(Message::Run(task));
+        Ok(())
+    }
+}
+
+impl PoolState {
+    fn send(&self, msg: Message) {
+        self.tx.lock().unwrap().send(msg).unwrap();
+    }
+
+    fn work(&self,
+            idx: usize,
+            after_start: Option<Arc<Fn(usize) + Send + Sync>>,
+            before_stop: Option<Arc<Fn(usize) + Send + Sync>>) {
+        let _scope = enter().unwrap();
+        after_start.map(|fun| fun(idx));
+        loop {
+            let msg = self.rx.lock().unwrap().recv().unwrap();
+            match msg {
+                Message::Run(r) => r.run(),
+                Message::Close => break,
+            }
+        }
+        before_stop.map(|fun| fun(idx));
+    }
+}
+
+impl Clone for ThreadPool {
+    fn clone(&self) -> ThreadPool {
+        self.state.cnt.fetch_add(1, Ordering::Relaxed);
+        ThreadPool { state: self.state.clone() }
+    }
+}
+
+impl Drop for ThreadPool {
+    fn drop(&mut self) {
+        if self.state.cnt.fetch_sub(1, Ordering::Relaxed) == 1 {
+            for _ in 0..self.state.size {
+                self.state.send(Message::Close);
+            }
+        }
+    }
+}
+
+impl ThreadPoolBuilder {
```

Comment:
It might be good to offer a `block_on` function here that works the same as the top-level `block_on` function, but uses the `ThreadPool` as the `Executor`, e.g. `let res = ThreadPool::new_num_cpus().block_on(main_future);`.

Reply:
I don't understand what `block_on` would mean in this context.

Reply:
@aturon Spawn the future on the thread pool, then block the current thread until the result of the future is available.

Reply:
I'm unclear on why that would be desirable, though.

Reply:
@aturon Isn't that what you'd want by default when running an application with a single `async_main` on a thread pool?

Reply:
I figured you'd instead do:

```rust
LocalPool::new()
    .run_until(async_main, &ThreadPool::new_num_cpus())
```